Various efforts bits of tinkering to get a stable build.

SamBarker opened the pull request.

SamBarker commented on junit5-extension/src/test/java/io/kroxylicious/testing/kafka/junit5ext/TemplateTest.java:
```diff
@@ -62,40 +65,48 @@ public void testMultipleClusterSizesWithAdminParameters(@DimensionMethodSource(v
 
+            // Then
+            int actualClusterSize = admin.describeCluster().nodes().get().size();
+            assertThat(actualClusterSize).as("Expected cluster to have %s nodes", cluster.getNumOfBrokers()).isEqualTo(cluster.getNumOfBrokers());
```
Gives a clearer error message when the cluster doesn't have the expected size. Which I think leads to clearer errors when the test fails.

SamBarker commented on pom.xml:
```diff
@@ -215,7 +234,7 @@
                     <version>1.8.0</version>
                     <configuration>
-                        <groups>java.,javax.,org.,com.</groups>
+                        <groups>java.,javax.,org.,com.,io.,*,io.kroxylicious</groups>
```
Aligns with kroxylicious project.

k-wall commented on impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java:
```diff
@@ -192,10 +192,17 @@ public void start() {
+                    }
+                    catch (KafkaException e) {
+                        LOGGER.log(System.Logger.Level.WARNING, "failed to start server due to: " + e.getMessage(), e);
+                        server.awaitShutdown();
```
Are you sure about that in the KRaft case? It looks to me like if the broker role brokers were to fail only those would shutdown.  I think calling `server.shutdown()` in your catch block would do no harm and keep the code easier to reason about.  

As the catch block in `kafka/server/KafkaServer.scala:565` catches, rethrows `Thowable`, what would you think about broadening your catch block to catch `Throwable`?

SamBarker replied:
I was calling `server.shutdown` in one version (locally) and took it out as superfluous but given you've had the same thoughts I'll add it back.

SamBarker replied:
Re throwable. I've never liked catching throwable as I don't believe we should catch error as there is nothing sensible we can do to recover. 

Having said that this is a test framework we could match Kafka and do so.

SamBarker replied:
fixed both in b759c0f7fa69da4a36c44c6160fb8d7c824e8e44

SamBarker replied:
Hmm with shutdown in place starting to see
```
java.lang.NullPointerException: Cannot invoke "org.apache.kafka.common.metrics.Metrics.sensor(String)" because "metrics" is null
	at kafka.network.SocketServer.<init>(SocketServer.scala:91) ~[kafka_2.13-3.4.0.jar:?]
	at kafka.server.ControllerServer.startup(ControllerServer.scala:179) ~[kafka_2.13-3.4.0.jar:?]
	at kafka.server.KafkaRaftServer.$anonfun$startup$1(KafkaRaftServer.scala:98) ~[kafka_2.13-3.4.0.jar:?]
	at kafka.server.KafkaRaftServer.$anonfun$startup$1$adapted(KafkaRaftServer.scala:98) ~[kafka_2.13-3.4.0.jar:?]
	at scala.Option.foreach(Option.scala:437) ~[scala-library-2.13.10.jar:?]
	at kafka.server.KafkaRaftServer.startup(KafkaRaftServer.scala:98) ~[kafka_2.13-3.4.0.jar:?]
	at io.kroxylicious.testing.kafka.invm.InVMKafkaCluster.lambda$start$1(InVMKafkaCluster.java:196) ~[classes/:?]
	at org.awaitility.core.CallableCondition$ConditionEvaluationWrapper.eval(CallableCondition.java:99) [awaitility-4.2.0.jar:?]
	at org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:248) [awaitility-4.2.0.jar:?]
	at org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:235) [awaitility-4.2.0.jar:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:317) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) [?:?]
	at java.lang.Thread.run(Thread.java:1589) [?:?]
	```
Which suggests restarting is not actually viable.

k-wall replied:
That might suggest a kafka lifecycling defect in KRaft.  The Kafka API suggests that start/stop/restart should be a supporrted use-case.  @showuon WDYT?

k-wall commented on impl/src/main/java/io/kroxylicious/testing/kafka/common/Utils.java:
```diff
@@ -38,7 +42,53 @@ private Utils() {
+     * @param timeUnit the time unit
+     * @param expectedBrokerCount the expected broker count
+     */
+    public static void awaitExpectedBrokerCountFromBootstrapServers(Map<String, Object> connectionConfig, int timeout, TimeUnit timeUnit, Integer expectedBrokerCount) {
```
awaitExpectedBrokerCountFromBootstrapServers is completely flawed.  My intent was to probe each broker individually, get their view of the cluster and make sure that the all the nodes could see all the others.  The implementation doesn't achieve that.

The admin client connects to bootstrap to get the broker topology and then reconnects to the nodes. The `org.apache.kafka.clients.admin.KafkaAdminClient#describeCluster` call uses a `LeastLoadedNodeProvider` to decide which node within the cluster receives the RPC, which is non deterministic.   The node provided used by describeCluster isn't pluggable so we can't switch to a ConstantNodeIdProvider.

SamBarker replied:
OHH bother that explains the need for the additional `await` in the test, well sot of. As that change fixed the test regardless of which version of the await I used. 

I need to spend some time futzing with logging tomorrow to better understand what state we are getting back.

SamBarker replied:
How about https://github.com/kroxylicious/kroxylicious-junit5-extension/pull/99/commits/e8d8c3bc7b358c05b921c51bcd8f0f3cf15c5565 ?

SamBarker commented on impl/src/main/java/io/kroxylicious/testing/kafka/common/ListeningSocketPreallocator.java:
```diff
@@ -41,6 +50,43 @@ public Stream<ServerSocket> preAllocateListeningSockets(int num) {
+        return random.ints(PORT_RANGE_LOW, PORT_RANGE_HIGH)
+                .mapToObj(number -> {
+                    try {
+                        var serverSocket = new ServerSocket(number);
```
@lift ignore 

We are just testing if the port is free, we don't actually accept connections using it.

SamBarker replied:
@sonatype-lift ignore
 
 We are only using the socket to check if the port is free we don't actually accept connections over it.

SamBarker replied:
@sonatype-lift ignore

SamBarker replied:
@sonatype-lift ignore

SamBarker commented on impl/src/main/java/io/kroxylicious/testing/kafka/common/ListeningSocketPreallocator.java:
```diff
@@ -23,16 +27,21 @@
+    private static final int PORT_RANGE_HIGH = 30_000;
     private final List<ServerSocket> all = new ArrayList<>();
 
+    private final Random random = new Random();
```
@sonatype-lift ignore

Are picking random numbers in arange for port allocation so we don't need cryptographically secure number generation.

SamBarker replied:
@sonatype-lift ignore

SamBarker commented on impl/src/main/java/io/kroxylicious/testing/kafka/common/ListeningSocketPreallocator.java:
```diff
@@ -7,39 +7,88 @@
+     * @param num number of ports to pre-allocate
+     * @return a List of bound sockets
+     */
+    public List<ServerSocket> preAllocateListeningSockets(int num) {
```
Not convinced by returning a list here as the allocator doesn't care about the order, but every invocation of the stream version turned it into a list (as they use index based access) so it seemed like the expedient choice here.

SamBarker commented on impl/src/main/java/io/kroxylicious/testing/kafka/common/ListeningSocketPreallocator.java:
```diff
@@ -7,39 +7,88 @@
+        if (num < 1) {
+            return List.of();
+        }
+        if (all.size() + num > MAX_PORT_COUNT) {
+            throw new IllegalArgumentException("Can't request more than " + MAX_PORT_COUNT + " ports. There are already " + all.size() + " allocated.");
+        }
```
Turns out not to be an effective guard as we hit open socket limits long before this point. I'm minded to remove it.

Any objections?

k-wall replied:
Happy to see it go.

SamBarker commented on impl/src/main/java/io/kroxylicious/testing/kafka/common/ListeningSocketPreallocator.java:
```diff
@@ -7,39 +7,88 @@
+                    }
+                    return null;
+                }).filter(Objects::nonNull)
+                .peek(all::add)
```
`peek` is a bit of a controversial choice the java doc suggests its really for debugging but intellj suggests it for usages such as this. 

The alternative is `map` which adds to all and returns the passed in socket.

k-wall replied:
I don't feel strongly.  This use case seems okay to me.  Maybe a short comment would help as you don't often see this Streams method used.

SamBarker commented on impl/src/main/java/io/kroxylicious/testing/kafka/common/ListeningSocketPreallocator.java:
```diff
@@ -7,39 +7,88 @@
      * @return stream of ephemeral ports
      */
-    public Stream<ServerSocket> preAllocateListeningSockets(int num) {
+    public List<ServerSocket> preAllocateEphemeralListeningSockets(int num) {
```
We don't need two competing implementations however I've kept this here for now while we settle on one.

SamBarker commented on impl/src/main/java/io/kroxylicious/testing/kafka/common/Utils.java:
```diff
@@ -32,13 +41,82 @@
+
+                        final Boolean isQuorate = promise.getNow(false);
+                        if (isQuorate) {
+                            admin.deleteTopics(Set.of(CONSISTENCY_TEST));
```
Fire and forget as its just being a nice citizen.

SamBarker commented on impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java:
```diff
@@ -173,8 +132,87 @@ private KafkaConfig buildBrokerConfig(KafkaClusterConfig.ConfigHolder c, Path te
     @Override
     public void start() {
-        if (zooFactory != null) {
+        // kraft mode: per-broker: 1 external port + 1 inter-broker port + 1 controller port + 1 anon port
```
Moved as part of an attempt to re-bind ports in the event of a failure which didn't really work. 

Saving the actual work of constructing servers etc until start is called has a slight appeal so I've left it here, however happy to move it back if there are opinions that think it should be done earlier.

k-wall commented on impl/pom.xml:
```diff
@@ -48,16 +48,6 @@
         <dependency>
             <groupId>org.apache.zookeeper</groupId>
             <artifactId>zookeeper</artifactId>
-            <exclusions>
```
Curious why you made this change?

SamBarker replied:
I moved the exclusion up to dependency management as looking at the dependency graph I was seeing both log4j 1.x and 2.x in the graph.

k-wall commented on impl/src/main/java/io/kroxylicious/testing/kafka/common/Utils.java:
```diff
@@ -32,13 +41,82 @@
+                                    else {
+                                        log.debug("Current topicDescriptions: {}", topicDescriptions);
+                                        var topicDescription = topicDescriptions.get(CONSISTENCY_TEST);
+                                        final long partitionLeaders = topicDescription.partitions().stream().map(TopicPartitionInfo::leader)
```
I understand your approach now: checking that the number of leaders == number of brokers.  That seems smart.  Should there be a distinct in the pipeline?

SamBarker replied:
TL;DR I'm going to switch to RF=expected broker count while keeping the logic to check the result of describeTopic.

I think a `distinct` is a good addition yes.

The partitionLeader pattern comes from the abnormal workloads testing done in a previous life. In which we saw intermittent failures where the topic was created but we couldn't publish to it. Which is noted in the [AdminClient#createTopics javadoc](https://kafka.apache.org/32/javadoc/org/apache/kafka/clients/admin/Admin.html#createTopics(java.util.Collection,org.apache.kafka.clients.admin.CreateTopicsOptions)) 
> It may take several seconds after [CreateTopicsResult](https://kafka.apache.org/32/javadoc/org/apache/kafka/clients/admin/CreateTopicsResult.html) returns success for all the brokers to become aware that the topics have been created.

The fix in that case was to ensure each partition had a leader. In that testing scenario we cared about the number of partitions and the distribution of them which might be why we ended up doing more than just asserting the number of replicas. In our current case we would still need to check the result of describeTopic.

Poking at the describe topic results suggests the following would also work:
```java
final long partitionLeaders = topicDescription.partitions()
        .stream()
        .map(TopicPartitionInfo::replicas)
        .flatMap(List::stream)
        .filter(Objects::nonNull)
        .distinct()
        .count();
```

You're making a good point about the topic creation failing, we probably need to retry that too.

k-wall replied:
Sounds good.

k-wall commented on impl/src/main/java/io/kroxylicious/testing/kafka/common/Utils.java:
```diff
@@ -32,13 +41,82 @@
+     * @param expectedBrokerCount the expected broker count
+     */
+    public static void awaitExpectedBrokerCountInClusterViaTopic(Map<String, Object> connectionConfig, int timeout, TimeUnit timeUnit, Integer expectedBrokerCount) {
+        try (var admin = Admin.create(connectionConfig)) {
```
I think a couple of sentences explaining how we check that all brokers are present would be good.

k-wall commented on impl/src/main/java/io/kroxylicious/testing/kafka/common/Utils.java:
```diff
@@ -32,13 +41,82 @@
  */
 public class Utils {
     private static final Logger log = getLogger(Utils.class);
+    private static final String CONSISTENCY_TEST = "__consistencyTest";
```
```suggestion
    private static final String CONSISTENCY_TEST = "__org_kroxylicious_testing_consistencyTest";
```

SamBarker commented on impl/src/main/java/io/kroxylicious/testing/kafka/common/Utils.java:
```diff
@@ -32,71 +35,101 @@
+                    if (throwable != null) {
+                        log.warn("Failed to create topic: {} due to {}", CONSISTENCY_TEST, throwable.getMessage(), throwable);
+                        if (throwable instanceof RetriableException) {
+                            CompletableFuture.supplyAsync(() -> createTopic(expectedBrokerCount, admin));
```
@sonatype-lift ignore

SamBarker commented on impl/src/main/java/io/kroxylicious/testing/kafka/common/Utils.java:
```diff
@@ -32,71 +35,101 @@
+            log.debug("Creating topic: {} via {}", CONSISTENCY_TEST, connectionConfig.get(BOOTSTRAP_SERVERS_CONFIG));
+            // Note we don't wait for the Topic to be created, we assume it will complete eventually.
+            // As the thing we care about is it actually being replicated to the other brokers anyway so we wait to confirm replication.
+            createTopic(expectedBrokerCount, admin);
```
@sonatype-lift ignore

SamBarker commented:
> I'm still seeing a failing test:
> 
> ```
> [INFO] Running io.kroxylicious.testing.kafka.junit5ext.TemplateTest$Versions
> [ERROR] Tests run: 8, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 6.058 s <<< FAILURE! - in io.kroxylicious.testing.kafka.junit5ext.TemplateTest$Versions
> [ERROR] io.kroxylicious.testing.kafka.junit5ext.TemplateTest.testMultipleClusterSizes(KafkaCluster)[2]  Time elapsed: 1.574 s  <<< FAILURE!
> org.opentest4j.AssertionFailedError: expected: <2> but was: <3>
> ```

I see something similar intermittently:
```
Formatting /var/folders/1g/2nsqtt916vg8f5bftn1tp0580000gn/T/kafka12506146268283241654/broker-0 with metadata.version 3.3-IV0.
Formatting /var/folders/1g/2nsqtt916vg8f5bftn1tp0580000gn/T/kafka12506146268283241654/broker-1 with metadata.version 3.3-IV0.
Formatting /var/folders/1g/2nsqtt916vg8f5bftn1tp0580000gn/T/kafka12506146268283241654/broker-2 with metadata.version 3.3-IV0.
2023-05-10 16:34:42 DEBUG io.kroxylicious.testing.kafka.common.Utils:57 - Creating topic: __org_kroxylicious_testing_consistencyTest via localhost:15391,localhost:13109,localhost:19406
2023-05-10 16:34:42 DEBUG io.kroxylicious.testing.kafka.common.Utils:61 - Waiting for __org_kroxylicious_testing_consistencyTest to be replicated to 3 brokers
2023-05-10 16:34:42 DEBUG io.kroxylicious.testing.kafka.common.Utils:64 - Calling describe topic
2023-05-10 16:34:42 DEBUG io.kroxylicious.testing.kafka.common.Utils:71 - Cluster quorum test topic (__org_kroxylicious_testing_consistencyTest) doesn't exist yet
2023-05-10 16:34:43 DEBUG io.kroxylicious.testing.kafka.common.Utils:115 - Create topic future completed.
2023-05-10 16:34:43 DEBUG io.kroxylicious.testing.kafka.common.Utils:64 - Calling describe topic
2023-05-10 16:34:43 DEBUG io.kroxylicious.testing.kafka.common.Utils:79 - Current topicDescriptions: {__org_kroxylicious_testing_consistencyTest=(name=__org_kroxylicious_testing_consistencyTest, internal=false, partitions=(partition=0, leader=localhost:19406 (id: 2 rack: null), replicas=localhost:19406 (id: 2 rack: null), localhost:15391 (id: 0 rack: null), localhost:13109 (id: 1 rack: null), isr=localhost:19406 (id: 2 rack: null), localhost:15391 (id: 0 rack: null), localhost:13109 (id: 1 rack: null)), authorizedOperations=null)}
2023-05-10 16:34:43 WARN  kafka.server.ReplicaFetcherThread:72 - [ReplicaFetcher replicaId=0, leaderId=2, fetcherId=0] Error in response for fetch request (type=FetchRequest, replicaId=0, maxWait=500, minBytes=1, maxBytes=10485760, fetchData={__org_kroxylicious_testing_consistencyTest-0=PartitionData(topicId=UCCxvIb6Rk6srX9UxQRKZQ, fetchOffset=0, logStartOffset=0, maxBytes=1048576, currentLeaderEpoch=Optional[0], lastFetchedEpoch=Optional.empty)}, isolationLevel=READ_UNCOMMITTED, removed=, replaced=, metadata=(sessionId=INVALID, epoch=INITIAL), rackId=)
java.io.IOException: Connection to 2 was disconnected before the response was read
	at org.apache.kafka.clients.NetworkClientUtils.sendAndReceive(NetworkClientUtils.java:99) ~[kafka-clients-3.4.0.jar:?]
	at kafka.server.BrokerBlockingSender.sendRequest(BrokerBlockingSender.scala:113) ~[kafka_2.13-3.4.0.jar:?]
	at kafka.server.RemoteLeaderEndPoint.fetch(RemoteLeaderEndPoint.scala:78) ~[kafka_2.13-3.4.0.jar:?]
	at kafka.server.AbstractFetcherThread.processFetchRequest(AbstractFetcherThread.scala:309) ~[kafka_2.13-3.4.0.jar:?]
	at kafka.server.AbstractFetcherThread.$anonfun$maybeFetch$3(AbstractFetcherThread.scala:124) ~[kafka_2.13-3.4.0.jar:?]
	at kafka.server.AbstractFetcherThread.$anonfun$maybeFetch$3$adapted(AbstractFetcherThread.scala:123) ~[kafka_2.13-3.4.0.jar:?]
	at scala.Option.foreach(Option.scala:437) [scala-library-2.13.10.jar:?]
	at kafka.server.AbstractFetcherThread.maybeFetch(AbstractFetcherThread.scala:123) [kafka_2.13-3.4.0.jar:?]
	at kafka.server.AbstractFetcherThread.doWork(AbstractFetcherThread.scala:106) [kafka_2.13-3.4.0.jar:?]
	at kafka.server.ReplicaFetcherThread.doWork(ReplicaFetcherThread.scala:97) [kafka_2.13-3.4.0.jar:?]
	at kafka.utils.ShutdownableThread.run(ShutdownableThread.scala:96) [kafka_2.13-3.4.0.jar:?]
2023-05-10 16:34:43 WARN  kafka.server.ReplicaFetcherThread:72 - [ReplicaFetcher replicaId=1, leaderId=2, fetcherId=0] Error in response for fetch request (type=FetchRequest, replicaId=1, maxWait=500, minBytes=1, maxBytes=10485760, fetchData={__org_kroxylicious_testing_consistencyTest-0=PartitionData(topicId=UCCxvIb6Rk6srX9UxQRKZQ, fetchOffset=0, logStartOffset=0, maxBytes=1048576, currentLeaderEpoch=Optional[0], lastFetchedEpoch=Optional.empty)}, isolationLevel=READ_UNCOMMITTED, removed=, replaced=, metadata=(sessionId=INVALID, epoch=INITIAL), rackId=)
java.io.IOException: Connection to 2 was disconnected before the response was read
	at org.apache.kafka.clients.NetworkClientUtils.sendAndReceive(NetworkClientUtils.java:99) ~[kafka-clients-3.4.0.jar:?]
	at kafka.server.BrokerBlockingSender.sendRequest(BrokerBlockingSender.scala:113) ~[kafka_2.13-3.4.0.jar:?]
	at kafka.server.RemoteLeaderEndPoint.fetch(RemoteLeaderEndPoint.scala:78) ~[kafka_2.13-3.4.0.jar:?]
	at kafka.server.AbstractFetcherThread.processFetchRequest(AbstractFetcherThread.scala:309) ~[kafka_2.13-3.4.0.jar:?]
	at kafka.server.AbstractFetcherThread.$anonfun$maybeFetch$3(AbstractFetcherThread.scala:124) ~[kafka_2.13-3.4.0.jar:?]
	at kafka.server.AbstractFetcherThread.$anonfun$maybeFetch$3$adapted(AbstractFetcherThread.scala:123) ~[kafka_2.13-3.4.0.jar:?]
	at scala.Option.foreach(Option.scala:437) [scala-library-2.13.10.jar:?]
	at kafka.server.AbstractFetcherThread.maybeFetch(AbstractFetcherThread.scala:123) [kafka_2.13-3.4.0.jar:?]
	at kafka.server.AbstractFetcherThread.doWork(AbstractFetcherThread.scala:106) [kafka_2.13-3.4.0.jar:?]
	at kafka.server.ReplicaFetcherThread.doWork(ReplicaFetcherThread.scala:97) [kafka_2.13-3.4.0.jar:?]
	at kafka.utils.ShutdownableThread.run(ShutdownableThread.scala:96) [kafka_2.13-3.4.0.jar:?]
2023-05-10 16:34:44 WARN  org.apache.kafka.clients.NetworkClient:775 - [RaftManager nodeId=1] Connection to node 0 (localhost/127.0.0.1:24926) could not be established. Broker may not be available.
2023-05-10 16:34:44 WARN  org.apache.kafka.clients.NetworkClient:775 - [RaftManager nodeId=2] Connection to node 0 (localhost/127.0.0.1:24926) could not be established. Broker may not be available.

org.opentest4j.AssertionFailedError: 
Expected :2
Actual   :3
<Click to see difference>
```

I think from those stack traces we had three brokers but one of them fell over silently...

k-wall reviewed (approved):
Looks like a useful improvement to me.

SamBarker commented on impl/src/main/java/io/kroxylicious/testing/kafka/common/KafkaClusterConfig.java:
```diff
@@ -184,139 +184,145 @@ public Stream<ConfigHolder> getBrokerConfigs(Supplier<KafkaEndpoints> endPointCo
-                    saslPairs.append(value);
-                    saslPairs.append(" ");
-                });
+            Optional.of(users).orElse(Map.of()).forEach((key, value) -> {
```
@sonatype-lift ignore

SamBarker commented:
Merging this now as I see the same thing as @k-wall build stability is greatly improved but still not 100% consistent.
